src/decoder.rs

```diff
@@ -71,14 +69,12 @@ impl<'a> Decoder<'a> {
 
     /// read a `N` bytes into slice
     pub fn read<const N: usize>(&mut self) -> DecodeResult<[u8; N]> {
-        let end = self.index.checked_add(N).ok_or(DecodeError::AddOverflow)?;
-        let bytes = self
-            .buffer
-            .get(self.index..end)
-            .ok_or(DecodeError::EndOfBuffer { index: end })?
-            .try_into()?;
-        self.index = end;
-        Ok(bytes)
+        if N > self.buffer.len() {
+            return Err(DecodeError::NotEnoughBytes);
+        }
+        let (slice, remaining) = self.buffer.split_at(N);
+        self.buffer = remaining;
+        Ok(slice.try_into().unwrap())
```

Comment:
It's not possible for this `unwrap` to panic because the error type here is `Infallible` correct? I wouldn't mind a comment stating as much. Also, if the error is infallible, does that mean `into()` works?

Reply:
It's not possible for this to panic because the condition that could cause the panic is guaranteed by the previous code. This is the same optimization as removes the panic from the `split_at` call. _In general_, you can't rely on rustc removing panics, but this pattern in particular is very reliable.

Here's a little demo: https://godbolt.org/z/r7MhdfW51
If you break the code by checking `N-1` or such, you'll be able to see the panic or bounds_check_fail call appear.